### PR TITLE
Add ability to create standalone binaries with qjs

### PR DESCRIPTION
**Pull request by saghul**
Ref: https://github.com/quickjs-ng/quickjs/issues/438
Closes: https://github.com/quickjs-ng/quickjs/pull/441

### CODE REVIEW

File: qjs.c [hunk 1]
```diff
@@ -344,11 +399,15 @@ int main(int argc, char **argv)
 {
     JSRuntime *rt;
     JSContext *ctx;
-    JSValue ret;
+    JSValue ret = JS_UNDEFINED;
```
Comment (saghul) [inline qjs.c:404]: Unnecessary, but gcc on mingw thinks there a use before it's initialized otherwise 🤷

File: qjs.c [hunk 2]
```diff
@@ -45,11 +45,63 @@
...
+static BOOL is_standalone(const char *exe)
+{
+    FILE *exe_f = fopen(exe, "rb");
+    if (!exe_f)
+        return FALSE;
+    if (fseek(exe_f, -trailer_size, SEEK_END) < 0)
+        return FALSE;
+    uint8_t buf[TRAILER_SIZE];
+    if (fread(buf, 1, trailer_size, exe_f) != trailer_size)
+        return FALSE;
```
Comment (bnoordhuis) [inline qjs.c]: The returns leak `exe_f`

File: qjs.c [hunk 3]
```diff
@@ -482,6 +545,39 @@ int main(int argc, char **argv)
                 stack_size = parse_limit(opt_arg);
                 break;
             }
+            if (opt == 'c' || !strcmp(longopt, "compile")) {
+                if (!opt_arg) {
+                    if (optind >= argc) {
+                        fprintf(stderr, "qjs: missing file for -c\n");
+                        exit(2);
```
Comment (bnoordhuis) [inline qjs.c]: For my curiosity: why 2 as the exit code?

File: standalone.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  dw.setUint32(0, exeSize, true /* little-endian */);
+
+  // We use os.open() so we can set the permissions mask.
+  const newFd = os.open(outFile, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o755);
+
+  if (newFd < 0) {
+    throw new Error(`failed to create ${outFile}`);
+  }
+  if (os.write(newFd, newBuffer, 0, newBuffer.byteLength) < 0) {
+    throw new Error(`failed to write to output file`);
```
Comment (bnoordhuis) [inline standalone.js]: The throws leak `newFd`

File: standalone.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  const bytecode = new Uint8Array(offset - Trailer.Size);
+
+  r = exe.seek(offset, std.SEEK_SET);
+  if (r < 0) {
+    throw new Error(`seek error: ${-r}`);
+  }
+
+  exe.read(bytecode.buffer, 0, bytecode.length);
+  if (exe.error()) {
+    throw new Error('read error');
```
Comment (bnoordhuis) [inline standalone.js]: Ditto

**Review by bnoordhuis — APPROVED** [review]

**Comment by saghul** [comment]
Thanks for the review! I'll address the comments and update the docs.

File: qjs.c (hunk 3 above)
Comment (saghul) [reply qjs.c]: C&P, I'm changing it to 1 like the other exit codes. There are 2 left that exit with code 2, when we fail to allocate the runtime or context. Do we care about them being different?

**Pull request merged**